stats/awscloudwatch/collector.go

```diff
@@ -0,0 +1,102 @@
+package awscloudwatch
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/stats"
+	log "github.com/sirupsen/logrus"
+)
+
+// Collector collects metrics and send them to CloudWatch
+type Collector struct {
+	client          cloudWatchClient
+	bufferedSamples []*sample
+	bufferLock      sync.Mutex
+}
+
+// New returns a new Collector
+func New(clientFactory func() (cloudWatchClient, error)) (*Collector, error) {
+	client, err := clientFactory()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Collector{client: client}, nil
+}
+
+// Init does nothing, as the client is injected in New constructor
+func (c *Collector) Init() error {
+	return nil
+}
+
+// Run send collected metrics every second and when context is done
+func (c *Collector) Run(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-ticker.C:
+			c.reportMetrics()
+		case <-ctx.Done():
+			c.reportMetrics()
+			return
+		}
+	}
+}
+
+func (c *Collector) reportMetrics() {
+	c.bufferLock.Lock()
+	err := c.client.reportSamples(c.bufferedSamples)
```

Comment:
This is a slow call that makes network requests, it should be outside of the `bufferLock.Lock()` zone, so that it doesn't unnecessarily block `Collect()` while it's executing. Instead, you should do something like this: https://github.com/loadimpact/k6/blob/2a2fe2cc665e0d2b818c4f3ca7ce4fc9a5821294/stats/influxdb/collector.go#L101-L104